examples/ex18.hpp

```diff
@@ -1,490 +1,278 @@
+//
+// This code provide example problems for the Euler equations and implements
+// the time-dependent DG operator given by the equation:
+//
+//            (u_t, v)_T - (F(u), ∇ v)_T + <F̂(u, n), [[v]]>_F = 0.
+//
+// This operator is designed for explicit time stepping methods. Specifically, the
+// function DGHyperbolicConservationLaws::Mult implements the following transformation:
+//
+//                             u ↦ M⁻¹(-DF(u) + NF(u))
+//
+// where M is the mass matrix, DF is the weak divergence of flux, and NF is the
+// interface flux. The inverse of the mass matrix is computed element-wise by
+// leveraging the block-diagonal structure of the DG mass matrix. Additionally,
+// the flux-related terms are computed using the HyperbolicFormIntegrator.
+//
+// The maximum characteristic speed is determined for each time step. For more details,
+// refer to the documentation of DGHyperbolicConservationLaws::Mult.
+//
+
+#include <functional>
 #include "mfem.hpp"
 
-using namespace std;
-using namespace mfem;
-
-// Problem definition
-extern int problem;
-
-// Maximum characteristic speed (updated by integrators)
-extern double max_char_speed;
-
-extern const int num_equation;
-extern const double specific_heat_ratio;
-extern const double gas_constant;
+namespace mfem
+{
 
-// Time-dependent operator for the right-hand side of the ODE representing the
-// DG weak form.
-class FE_Evolution : public TimeDependentOperator
+/// @brief Time dependent DG operator for hyperbolic conservation laws
+class DGHyperbolicConservationLaws : public TimeDependentOperator
 {
 private:
-   const int dim;
-
-   FiniteElementSpace &vfes;
-   Operator &A;
-   SparseMatrix &Aflux;
-   DenseTensor Me_inv;
-
-   mutable Vector state;
-   mutable DenseMatrix f;
-   mutable DenseTensor flux;
+   const int dim; // domain dimension
+   const int num_equations; // the number of equations
+   FiniteElementSpace &vfes; // vector finite element space
+   // Element integration form. Should contain ComputeFlux
+   std::unique_ptr<HyperbolicFormIntegrator> formIntegrator;
+   // Base Nonlinear Form
+   std::unique_ptr<NonlinearForm> nonlinearForm;
+   // element-wise inverse mass matrix
+   std::vector<DenseMatrix> Me_inv;
+   // global maximum characteristic speed. Updated by form integrators
+   mutable double max_char_speed;
+   // auxiliary variable used in Mult
    mutable Vector z;
 
-   void GetFlux(const DenseMatrix &state_, DenseTensor &flux_) const;
+   // Compute element-wise inverse mass matrix
+   void ComputeInvMass();
 
 public:
-   FE_Evolution(FiniteElementSpace &vfes_,
-                Operator &A_, SparseMatrix &Aflux_);
+   /**
+    * @brief Construct a new DGHyperbolicConservationLaws object
+    *
+    * @param vfes_ vector finite element space. Only tested for DG [Pₚ]ⁿ
+    * @param formIntegrator_ integrator (F(u,x), grad v)
+    */
+   DGHyperbolicConservationLaws(
+      FiniteElementSpace &vfes_,
+      HyperbolicFormIntegrator *formIntegrator_);
```

Comment:
```suggestion
      std::unique_ptr<HyperbolicFormIntegrator> formIntegrator_);
```
Would it be better to make this change so it is clear from the constructor signature that this class will take ownership of the `formIntegrator_` (i.e., that this class will delete the memory pointed to by `formIntegrator_`)?  I'll note there may need to be other changes to be consistent here, such as
- using `std::make_unique<HyperbolicFormIntegrator>` instead of `new HyperbolicFormIntegrator` in the calling code
- possibly needing `formIntegrator(std::move(formIntegrator_))` instead of `formIntegrator(formIntegrator_)`

Apologies for the staggered comments, I just noticed this as I'm preparing to present this PR to a team working on hyperbolic problems

Reply:
Thank you for your suggestion. I'll make changes in the constructor. However, `std::make_unique` was introduced in C++14, and I'm currently using C++11. C++11 is required since MFEM 4.0, but there is no requirements for C++14 in `INSTALL`. I think we should keep `reset()` until MFEM requires C++14 or higher. I'm not that familiar with using `unique_ptr` (actually, this is the first time that I used `std::unique_ptr`). So, if there is a better way to initialize the `unique_ptr` other than the current way in C++11, please let me know.

Reply:
Ah good point.  One can just use the `std::unique_ptr<T>` constructor wrapped around the `new T` call:
`std:unique_ptr<HyperbolicFormIntegrator>(new HyperbolicFormIntegrator())` instead of `std::make_unique<HyperbolicFormIntegrator>()`... or maybe the compile will already do things correctly leaving the `new HyperbolicFormIntegrator` call as-is.

Reply:
Thanks for the suggestion @cjvogl (and @dohyun-cse for making the change).

Personally I'm a little conflicted about this (especially since `make_unique` isn't available in C++11 (though it would be trivial to add our own version of it, it doesn't do anything complicated)). It just adds some verbosity (but I agree the intent is more clear). Also FYI you can't leave only the `new HyperbolicFormIntegrator` since the `T*` constructor of `unique_ptr` is `explicit`.

Since this is just in ex18.hpp and not in the main library, I don't think we need to worry too much about it, so either way is probably fine.

If we really want to get into over-engineering/bike-shedding, we can do something like this:

```cpp
auto euler = DGHyperbolicConservationLaws::New<HyperbolicFormIntegrator>(vfes, numericalFlux, IntOrderOffset);
```

so the integrator type becomes a template parameter, and the arguments to its constructor are passed after the FE space.

The function would be defined like this:

```cpp
template <typename T, typename... Args>
static DGHyperbolicConservationLaws New(FiniteElementSpace &vfes_, Args... args)
{
   T *t = nullptr;
   return DGHyperbolicConservationLaws(vfes_, t, std::forward<Args>(args)...);
}
```

and the constructor would be made private and defined like this:

```cpp
template <typename T, typename... Args>
DGHyperbolicConservationLaws::DGHyperbolicConservationLaws(
   FiniteElementSpace &vfes_, T *t, Args... args)
   : TimeDependentOperator(vfes_.GetTrueVSize()),
     formIntegrator(new T(std::forward<Args>(args)...)),
     num_equations(formIntegrator->num_equations),
     vfes(vfes_),
     z(vfes_.GetTrueVSize())
{ […] }
```

(You need to use a "named constructor" because the only way to specify a template parameter for an actual constructor is through template argument deduction, which is why you need the `T *t = nullptr` dummy value).

Reply:
> Also FYI you can't leave only the new HyperbolicFormIntegrator since the T* constructor of unique_ptr is explicit.

Good to know, thanks for the heads up!

> Since this is just in ex18.hpp and not in the main library, I don't think we need to worry too much about it, so either way is probably fine.

I see your point... FWIW, I was concerned about the perhaps improbable future where I point my collaborators to example 18, they reproduce the `DGHyperbolicConservationLaws` class in their code, and then (for whatever reason) pass a `HyperbolicFormIntegrator` object by reference to the constructor.

Reply:
It seems to me that it is getting over-complicated. I think passing/storing `HyperbolicFormIntegrator` as a reference is the easiest way (we will not pass `nullptr`. So no harm done). This makes the ownership clear, and the constructor is easier to use. Since it is an example code, I think readability should come first.

According to the guideline in `CONTRIBUTING.md` (line 349), we can just revert it to the previous version with a raw pointer and add comment about the ownership. But I think it is error-prone (who reads the documentation before you get segfault?) So I like the first approach with reference.

Reply:
>  I think passing/storing HyperbolicFormIntegrator as a reference is the easiest way (we will not pass nullptr. So no harm done).

This has two disadvantages from my perspective:
- it requires the calling code to own/maintain the object instead of transferring ownership
- it breaks precedence from all the other similar "form" implementations (e.g., `BilinearForm`) that assume the ownership of the integrators is transferred by default

> But I think it is error-prone (who reads the documentation before you get segfault?) 

I agree.

I recommend just leaving the code as it currently is.  Is one line more verbose than it maybe needs to be, sure.... but I personally prefer paying the cost in verbosity in one line for the leveraging of the `std::unique_ptr` class and the clear move semantics that come with it.  Once the debate between which C++ standard to move to is decided (i.e., between c++14 vs c++17), the code will be able to use `std::make_unique` anyway

Reply:
That's fine with me. I also don't want to keep changing the code after I got approvals. I'll leave it as it is. If future MFEM supports C++14 or higher, then we can create another issue/PR to discuss further.